### PR TITLE
codegen: Split up source, header and Declarations.yaml generation

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67497
* #67496

This allows more of the code-generation to happen in parallel, whereas
previously all codegen was serialized.

Differential Revision: [D32027250](https://our.internmc.facebook.com/intern/diff/D32027250)

### CODE REVIEW

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -63,6 +63,7 @@ if(INTERN_BUILD_ATEN_OPS)
   set(CMAKE_POSITION_INDEPENDENT_CODE ${__caffe2_CMAKE_POSITION_INDEPENDENT_CODE})
 
   # Generate the headers wrapped by our operator
+  file(GLOB_RECURSE all_python "${PROJECT_SOURCE_DIR}/tools/codegen/*.py")
```
Comment (albanD) [inline caffe2/CMakeLists.txt:66]: That doesn't contain the autograd codegen? Is that on pupose?
Also this folder is nested, should this be `/**/*.py`?

**Review by albanD — COMMENTED** [review]
The code moved and update in CMake looks good.
Is there a way to test the dependency graph for the build to make sure we didn't miss something?

FYI @bdhirsh who works on this code

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32027250).

File: caffe2/CMakeLists.txt (hunk 1 above)
Comment (peterbell10) [reply caffe2/CMakeLists.txt:66]: >That doesn't contain the autograd codegen? Is that on pupose?

Yes, the generator only actually uses `CodeTemplate` which is in `tools/codegen/`.  
https://github.com/pytorch/pytorch/blob/master/caffe2/contrib/aten/gen_op.py

>Also this folder is nested, should this be /**/*.py?

`ninja` confirms this includes recursive directories.
```
$ ninja -C build -t query /home/peter/git/pytorch/build/caffe2/contrib/aten/aten_op.h
/home/peter/git/pytorch/build/caffe2/contrib/aten/aten_op.h:
  input: CUSTOM_COMMAND
    ../tools/codegen/__init__.py
    ../tools/codegen/api/__init__.py
    ../tools/codegen/api/autograd.py
    ../tools/codegen/api/cpp.py
    ../tools/codegen/api/dispatcher.py
    ../tools/codegen/api/functionalization.py
    ../tools/codegen/api/meta.py
    ../tools/codegen/api/native.py
    ../tools/codegen/api/python.py
    ../tools/codegen/api/structured.py
    ../tools/codegen/api/translate.py
    ../tools/codegen/api/types.py
    ../tools/codegen/code_template.py
    ../tools/codegen/context.py
    ../tools/codegen/dest/__init__.py
    ../tools/codegen/dest/native_functions.py
    ../tools/codegen/dest/register_dispatch_key.py
    ../tools/codegen/gen.py
    ../tools/codegen/gen_backend_stubs.py
    ../tools/codegen/gen_functionalization_type.py
    ../tools/codegen/local.py
    ../tools/codegen/model.py
    ../tools/codegen/selective_build/__init__.py
    ../tools/codegen/selective_build/operator.py
    ../tools/codegen/selective_build/selector.py
    ../tools/codegen/utils.py
    aten/src/ATen/Declarations.yaml
    ../caffe2/contrib/aten/gen_op.py
    ../caffe2/contrib/aten/aten_op_template.h
  outputs:
```

**Comment by peterbell10** [comment]
>Is there a way to test the dependency graph for the build to make sure we didn't miss something?

You can use `ninja -C build -t targets all` to list all build targets, then `ninja -C build -t query [target]` prints the dependencies for that target.

**Review by albanD — APPROVED** [review]
Sounds good! Thanks.

**Pull request closed without merging**